src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java

```diff
@@ -0,0 +1,328 @@
+ * this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
+ */
+
+package com.mysql.cj.jdbc.ha.plugins;
+
+import com.mysql.cj.Messages;
+import com.mysql.cj.conf.ConnectionUrl;
+import com.mysql.cj.conf.HostInfo;
+import com.mysql.cj.conf.PropertyKey;
+import com.mysql.cj.conf.PropertySet;
+import com.mysql.cj.conf.RuntimeProperty;
+import com.mysql.cj.exceptions.MysqlErrorNumbers;
+import com.mysql.cj.jdbc.JdbcConnection;
+import com.mysql.cj.log.Log;
+
+import java.lang.reflect.Method;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+public class ReadWriteSplittingPlugin implements IConnectionPlugin {
+  public static final int WRITER_INDEX = 0;
+  public static final String METHOD_SET_READ_ONLY = "setReadOnly";
+
+  private final ICurrentConnectionProvider currentConnectionProvider;
+  private final IConnectionProvider connectionProvider;
+  private final PropertySet propertySet;
+  private final IConnectionPlugin nextPlugin;
+  private final Log logger;
+
+  private boolean inTransaction = false;
+  private boolean readOnly = false;
+  private List<HostInfo> hosts = new ArrayList<>();
+  private JdbcConnection writerConnection;
+  private JdbcConnection readerConnection;
+
+  public ReadWriteSplittingPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this(currentConnectionProvider, new BasicConnectionProvider(), propertySet, nextPlugin, logger);
+  }
+
+  ReadWriteSplittingPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      IConnectionProvider connectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this.currentConnectionProvider = currentConnectionProvider;
+    this.connectionProvider = connectionProvider;
+    this.propertySet = propertySet;
+    this.nextPlugin = nextPlugin;
+    this.logger = logger;
+  }
+
+  @Override
+  public Object executeOnConnectionBoundObject(Class<?> methodInvokeOn, String methodName, Callable<?> executeSqlFunc, Object[] args) throws Exception {
+    return this.nextPlugin.executeOnConnectionBoundObject(methodInvokeOn, methodName, executeSqlFunc, args);
+  }
+
+  @Override
+  public Object executeOnConnection(Method method, List<Object> args) throws Exception {
+    if (METHOD_SET_READ_ONLY.equals(method.getName()) && args != null && args.size() > 0) {
+      switchConnectionIfRequired((Boolean) args.get(0));
+    }
+    return this.nextPlugin.executeOnConnection(method, args);
+  }
+
+  void switchConnectionIfRequired(Boolean readOnly) throws SQLException {
+    if (readOnly == null) return;
+    final JdbcConnection currentConnection = this.currentConnectionProvider.getCurrentConnection();
+    if (readOnly) {
```

Comment:
`(!isReaderConnection() || currentConnection.isClosed())` is checked here and in `switchToReaderConnection`. Similarly for `(!isWriterConnection() || currentConnection.isClosed())` and `switchToWriterConnection`.
Is it possible to simply the conditions here to:
```java
  if (readOnly) {
    if (!this.inTransaction) {
      ...
    }
  } else {
    if (this.inTransaction) {
      ...
    }

    ...
  }
```

Reply:
The first one could work I think. If we changed it we would be entering switchToReaderConnection even when the current connection is the reader connection and it is open, in which case we don't need to switch connections, however switchToReaderConnection will detect this and return without doing anything. I don't think the second one would work though. If we changed the second check, it means that if a user was in a transaction on an open writer connection and called setReadOnly(false), even though we don't need to do anything, the error inside the inTransaction check would be thrown